modular_chomp/code/modules/pda/core_apps.dm

```diff
@@ -0,0 +1,163 @@
+		return TRUE
+
+	switch(action)
+		if("switch-to-onduty-rank")
+			if(checkFace())
+				if(checkCardCooldown())
+					makeOnDuty(params["switch-to-onduty-rank"], params["switch-to-onduty-assignment"])
+			return TRUE
+		if("switch-to-offduty")
+			if(checkFace())
+				if(checkCardCooldown())
+					makeOffDuty()
+			return TRUE
+
+/datum/data/pda/app/timeclock/proc/getOpenOnDutyJobs(var/mob/user, var/department)
+	var/list/available_jobs = list()
+	for(var/datum/job/job in job_master.occupations)
+		if(isOpenOnDutyJob(user, department, job))
+			available_jobs[job.title] = list(job.title)
+			if(job.alt_titles)
+				for(var/alt_job in job.alt_titles)
+					if(alt_job != job.title)
+						available_jobs[job.title] += alt_job
+	return available_jobs
+
+/datum/data/pda/app/timeclock/proc/isOpenOnDutyJob(var/mob/user, var/department, var/datum/job/job)
+	return job \
+		   && job.is_position_available() \
+		   && !job.whitelist_only \
+		   && !jobban_isbanned(user,job.title) \
+		   && job.player_old_enough(user.client) \
+		   && job.player_has_enough_playtime(user.client) \
+		   && job.pto_type == department \
+		   && !job.disallow_jobhop \
+		   && job.timeoff_factor > 0
+
+/datum/data/pda/app/timeclock/proc/makeOnDuty(var/newrank, var/newassignment)
+	var/datum/job/oldjob = job_master.GetJob(pda.id.rank)
+	var/datum/job/newjob = job_master.GetJob(newrank)
+	if(!oldjob || !isOpenOnDutyJob(pdaUser, oldjob.pto_type, newjob))
+		return
+	if(newassignment != newjob.title && !(newassignment in newjob.alt_titles))
+		return
+	if(newjob.camp_protection && round_duration_in_ds < CONFIG_GET(number/job_camp_time_limit))
+		if(SSjob.restricted_keys.len)
+			var/list/check = SSjob.restricted_keys[newjob.title]
+			if(pdaUser.client.ckey in check)
+				to_chat(pdaUser,span_danger("[newjob.title] is not presently selectable because you played as it last round. It will become available to you in [round((CONFIG_GET(number/job_camp_time_limit) - round_duration_in_ds) / 600)] minutes, if slots remain open."))
+				return
+
+	if(newjob)
+		newjob.register_shift_key(pdaUser.client.ckey)
+		pda.id.access = newjob.get_access()
+		pda.id.rank = newjob.title
+		pda.id.assignment = newassignment
+		pda.id.name = text("[pda.id.registered_name]'s ID Card ([pda.id.assignment])")
+		data_core.manifest_modify(pda.id.registered_name, pda.id.assignment, pda.id.rank)
+		pda.id.last_job_switch = world.time
+		callHook("reassign_employee", list(pda.id))
+		newjob.current_positions++
+		var/mob/living/carbon/human/H = pdaUser
+		H.mind.assigned_role = pda.id.rank
+		H.mind.role_alt_title = pda.id.assignment
+		announce.autosay("[pda.id.registered_name] has moved On-Duty as [pda.id.assignment].", "Employee Oversight", channel, zlevels = using_map.get_map_levels(get_z(src)))
+	return
+
+/datum/data/pda/app/timeclock/proc/makeOffDuty()
+	var/datum/job/foundjob = job_master.GetJob(pda.id.rank)
+	if(!foundjob)
+		return
+	if((get_area(pdaUser).flags & TIMECLOCK_RESTRICTED))
```

Comment:
flag_check

also, those need to be added to the areas as well

Furthermore, I would allow it always if isbelly(loc)

Reply:
The latter's simple enough, for adding flags to areas, should I just add on to the existing Space Station 13 areas_ch.dm file, or make a new one in modular? Or is there somewhere else I should do that?

Reply:
our areas are by map

modular_chomp then
souther_cross_areas.dm
soluna_nexus_areas.dm
relicbase_areas.dm

Reply:
also from how area flags work, I would turn this into AREA_ALLOW_CLOCKOUT

because most areas already have special flags, aka the bar, maints and dorms which can be appended easier than trying the inverted behaviour.

Reply:
Right, I'll do that, then, I've flagged the dorms, most clearly public areas, hallways, and some more distinct parts of maintenance to allow clocking out on all the maps, as well as the planet surface. I'll push the commit once I test to make sure it works as intended, let me know if any of the areas should be unflagged, or if there are any that should be flagged I missed, wouldn't be surprised if I missed a few, though I tried to double check each file. Already know I missed the SC cafeteria, holodeck and game room, because those aren't defined in southern_cross_areas.dm

Reply:
Be careful with flags, if you apply new flags to any child, the parent flags will not be applied, you ned to explicitly flag areas.

Reply:
I think all the flags I set are good. Looked over all of them and didn't see anything immediately being overridden.